lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -307,30 +312,62 @@ def from_string(cls, collection_input, artifacts_manager, supplemental_signature
 
     @classmethod
     def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_options=True):
-        req_name = collection_req.get('name', None)
-        req_version = collection_req.get('version', '*')
-        req_type = collection_req.get('type')
+        result = ArgumentSpecValidator(
+            argument_spec=dict(
+                name=dict(type='str', required=True),
+                signatures=dict(type='list', elements='str'),
+                source=dict(type='raw'),
+                type=dict(type='str'),
+                version=dict(default='*', type='str')
+            )
+        ).validate(collection_req)
+
+        req_name = result.validated_parameters.get('name')
+        req_version = result.validated_parameters.get('version')
+        req_type = result.validated_parameters.get('type')
         # TODO: decide how to deprecate the old src API behavior
-        req_source = collection_req.get('source', None)
-        req_signature_sources = collection_req.get('signatures', None)
+        req_source = result.validated_parameters.get('source')
+        req_signature_sources = result.validated_parameters.get('signatures')
+        if result.error_messages:
+            error = result.errors[0]
+            if isinstance(error, UnsupportedError):
+                raise AnsibleError(
+                    "The following {collection_name!s} collection requirement entry keys are not "
+                    "valid: {unsupported_params!s}".format(
+                        collection_name=req_name,
+                        unsupported_params=', '.join(sorted(result.unsupported_parameters))
+                    )
+                )
+            raise AnsibleError(
+                f"Failed to install collection requirement entry: {error.msg}"
+            )
```

Comment:
Rather than special-casing the first failure or error type, this should display all errors. I would also not mention `install` specifically, since there are multiple subcommands that use this method.

```suggestion
             raise AnsibleError(
                f"Failed to parse collection requirement entry: {', '.join(result.error_messages)}"
             )
```

If there's an issue with unsupported parameters being randomly ordered (not sure if there is), that could be fixed in the argument spec validator.

Reply:
Agreed on omitting the subcommand in the error message. I believe I special-cased the first error type because the error message returned wouldn't read the best when appended. For example, if the `foo` and `bar` keys are used as part of a requirement, the exception raised will have the message `ERROR! Failed to parse collection requirement entry: bar, foo. Supported parameters include: name, signatures, source, type, version.`.

Reply:
Good point, it doesn't really include context about which requirement specifies the unsupported keys.

How about `ERROR! Failed to parse collection requirement entry: {'name': 'amazon.aws', 'version': '6.0.0', 'foo': 'baz', 'bar': 'qux'}: bar, foo. Supported parameters include: name, signatures, source, type, version.`?

```python
        if result.error_messages:
             raise AnsibleError(
                f"Failed to parse collection requirement entry: {collection_req}: {', '.join(result.error_messages)}"
             )
```

Reply:
Yea, that'll work.